lib/jackalope/work_list.ex

```diff
@@ -42,6 +43,35 @@ defmodule Jackalope.WorkList do
     %__MODULE__{work_list | items: tl(work_list.items)}
   end
 
+  @spec pending(t(), reference()) :: t()
+  def pending(work_list, ref) do
+    item = hd(work_list.items)
+
+    %__MODULE__{
+      work_list
+      | items: tl(work_list.items),
+        pending: Map.put(work_list.pending, ref, item)
+    }
+  end
+
+  @spec reset_pending(Jackalope.WorkList.t()) :: Jackalope.WorkList.t()
+  def reset_pending(work_list) do
+    pending_items =
+      for(
+        {ref, item} when is_reference(ref) <- work_list.pending,
+        do: item
+      )
+
+    prepend(%__MODULE__{work_list | pending: {}}, pending_items)
+  end
+
+  @spec done(t(), reference()) :: {t(), any}
+  def done(work_list, ref) do
+    item = Map.get(work_list.pending, ref)
+    # item can be nil
+    {%__MODULE__{work_list | pending: Map.delete(work_list.pending, ref)}, item}
```

Comment:
```suggestion
    {item, pending} = Map.pop(work_list.pending, ref)
    # item can be nil
    {%__MODULE__{work_list | pending: pending}, item}
```

Reply:
All fixed and added tests for WorkList pending, done, reset_pending